### PR TITLE
Share state between client and mobile-client

### PR DESCRIPTION
**Pull request by sliptype**
### Description

This is the first step towards the promised land of sharing state & logic between clients. Lots of files in this PR but it's mostly moving dependencies around.

This PR:
* Creates a `/common` directory where logic shared between clients will live for now (will eventually be pulled into a separate package)
* Moves the state slices and hooks required for the BottomBar into `/common` in order to migrate the BottomBar to react native

Interesting files to look at include:
* `src/store/configureStore.js`
* `src/store/reducers.js`
* `src/common/store/index`

### Dragons
This moves a lot of dependencies around and adds a subscription to the redux store in native mobile. Could have an impact on performance so let's keep an eye on that

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide repro instructions & any configuration.

### How will this change be monitored?

For features that are critical or could fail silently please describe the monitoring/alerting being added.

### CODE REVIEW

**Comment by audius-infra** [comment]
Preview this change https://frontend-demo.audius.co/sk-client-store

**Comment by audius-infra** [comment]
Preview this change https://frontend-demo.audius.co/sk-client-store

**Comment by audius-infra** [comment]
Preview this change https://frontend-demo.audius.co/sk-client-store

**Comment by audius-infra** [comment]
Preview this change https://frontend-demo.audius.co/sk-client-store

**Comment by audius-infra** [comment]
Preview this change https://frontend-demo.audius.co/sk-client-store

**Comment by audius-infra** [comment]
Preview this change https://frontend-demo.audius.co/sk-client-store

**Comment by audius-infra** [comment]
Preview this change https://frontend-demo.audius.co/sk-client-store

**Comment by audius-infra** [comment]
Preview this change https://frontend-demo.audius.co/sk-client-store

**Comment by audius-infra** [comment]
Preview this change https://frontend-demo.audius.co/sk-client-store

**Review by jowlee — APPROVED** [review]

File: src/store/configureStore.js [hunk 1]
```diff
@@ -114,9 +120,39 @@ const middlewares = applyMiddleware(
 
 let store = null
 
+// As long as the mobile client is dependent on the web client, we need to sync
+// the client store from web -> mobile
+const clientStoreKeys = Object.keys(clientStoreReducers)
```
Comment (raymondjacobson) [inline src/store/configureStore.js:125]: not that this is super huge, but can we put this in a Set? would make the lookup on each subscription O(1)

File: src/store/configureStore.js [hunk 2]
```diff
@@ -114,9 +120,39 @@ const middlewares = applyMiddleware(
...
+
+    store.subscribe(() => {
+      const state = store.getState()
+      const previousState = currentState
+      currentState = state
+      if (
+        !previousState ||
+        clientStoreKeys.some(k => currentState[k] !== previousState[k])
+      ) {
+        // Debounce messages to minimize expensive stringify and parse.
```
Comment (raymondjacobson) [inline src/store/configureStore.js:140]: This debounce can be a lot smaller.

console.time('t');a = JSON.stringify(store.getState()); JSON.parse(a);console.timeEnd('t')

If you run this from the client in the console it's < 5ms. i believe mobile environments should be quite similar. not entirely sure what the overhead of the message pass is, but it's still not going to be 500ms

**Review by raymondjacobson — APPROVED** [review]
Couple little things

Also, should we be doing named exports from the new stuff, like common/models/Status

**Pull request merged**